src/NNLS.jl

```diff
@@ -534,4 +536,145 @@ function nnls{T}(A::DenseMatrix{T}, b::DenseVector{T}, max_iter=(3 * size(A, 2))
+# Variable names match the paper wherever possible
+
+@static if VERSION < v"0.6-"
+    typealias AllColsSubArray{T} SubArray{T,2,Array{T,2},Tuple{UnitRange{Int},Colon},false}
+else
+    const AllColsSubArray{T} = SubArray{T,2,Array{T,2},Tuple{UnitRange{Int},Base.Slice{Base.OneTo{Int}}},false}
+end
+
+type QPWorkspace{T<:LinAlg.BlasReal, I}
+    # Variables from paper:
+    L::Matrix{T}
+    c::Vector{T}
+    G::Matrix{T}
+    g::Vector{T}
+    M::Matrix{T}
+    d::Vector{T}
+    r::Vector{T}
+
+    # Additional variables:
+    e::Vector{T} # intermediate variable
+    A::Matrix{T} # 'A'-matrix of NNLS problem
+    AM::AllColsSubArray{T} # upper block of A
+    Ad::AllColsSubArray{T} # last row of A
+    b::Vector{T} # 'b'-vector of NNLS problem
+    nnlswork::NNLSWorkspace{T, I}
+    status::Symbol
+
+    function QPWorkspace(n::Integer, q::Integer)
+        L = Matrix{T}(n, n)
+        c = Vector{T}(n)
+        G = Matrix{T}(q, n)
+        g = Vector{T}(q)
+        M = Matrix{T}(q, n)
+        d = Vector{T}(q)
+        r = Vector{T}(n + 1)
+        e = Vector{T}(n)
+        A = Matrix{T}(n + 1, q)
+        AM = view(A, 1 : n, :)
+        Ad = view(A, n + 1 : n + 1, :)
+        b = Vector{T}(n + 1)
+        work = NNLSWorkspace{T, I}(size(A)...)
+        status = :Unsolved
+        new{T, I}(L, c, G, g, M, d, r, e, A, AM, Ad, b, work, status)
+    end
+end
+
+"""
+    load!(work::QPWorkspace, Q, c, G, g)
+
+Load problem data for the QP
+
+Minimize ``\\frac{1}{2} z' Q z + c' z``
+Subject to ``G z \\leq g``
+"""
+function load!{T}(work::QPWorkspace{T}, Q::AbstractMatrix{T}, c::AbstractVector{T}, G::AbstractMatrix{T}, g::AbstractVector{T})
+    work.L .= Q
+    work.c .= c
+    work.G .= G
+    work.g .= g
+    work.status = :Unsolved
+    nothing
+end
+
+# TODO: resize!?
+"""
+    solve!(work::QPWorkspace)
+
+Solve the QP that was loaded into `work` using load!.
+"""
+function solve!{T}(work::QPWorkspace{T}, eps_infeasible = 1e-4) # TODO: method name? Maybe change nnls! to solve! too?
+    work.status == :Unsolved || error("Problem was already solved.")
+
+    L = work.L
+    c = work.c
+    G = work.G
+    g = work.g
+    M = work.M
+    d = work.d
+    e = work.e
+    A = work.A
+    AM = work.AM
+    Ad = work.Ad
+    b = work.b
+    r = work.r
+    nnlswork = work.nnlswork
+
+    # Compute M
+    LinAlg.LAPACK.potrf!('U', L) # L <- upper cholesky factor of Q
+    M .= G
+    LinAlg.BLAS.trsm!('R', 'U', 'N', 'N', 1., L, M) # M <- G L⁻¹
+
+    # Compute d
+    e .= c
+    LinAlg.LAPACK.potrs!('U', L, e) # e <- Q⁻¹ c
+
+    d .= g
+    LinAlg.BLAS.gemv!('N', 1., G, e, 1., d) # d <- g + G Q⁻¹ c
+
+    # Populate A
+    transpose!(AM, M)
+    transpose!(Ad, d)
+    scale!(A, -1)
+
+    # Populate b
+    γ = one(T)
+    b[:] = 0
+    b[end] = γ
+
+    # Solve the NNLS
+    load!(nnlswork, A, b)
+    nnls!(nnlswork)
+    y = nnlswork.x
+
+    # Compute the residual
+    r = b
+    LinAlg.BLAS.gemv!('N', 1., A, y, -1., r) # r <- A * y - b
```

Comment:
It might be possible to clean this up later using https://github.com/simonbyrne/InplaceOps.jl but, again, I'm happy with the code as written. It's not like my NNLS code is particularly readable ;-)